packages/@n8n/nodes-langchain/nodes/vendors/Microsoft/microsoft-utils.ts

```diff
@@ -0,0 +1,282 @@
+
+	return async (turnContext: TurnContext) => {
+		const agentId = turnContext.activity.recipient?.agenticAppId ?? clientId;
+		const agentName = turnContext.activity.recipient?.name ?? 'Microsoft Agent 365';
+		const tenantDetails: TenantDetails = {
+			tenantId: turnContext.activity.recipient?.tenantId ?? tenantId ?? '',
+		};
+		const conversationId = turnContext.activity.conversation?.id;
+		const inputText = turnContext.activity.text || '';
+
+		const baggageScope = new BaggageBuilder()
+			.tenantId(tenantDetails.tenantId)
+			.agentId(agentId)
+			.correlationId(uuid())
+			.agentName(agentName)
+			.conversationId(conversationId)
+			.build();
+
+		await baggageScope.run(async () => {
+			const invokeAgentDetails: InvokeAgentDetails = {
+				agentId,
+				agentName,
+				conversationId,
+				request: {
+					content: inputText || 'Unknown text',
+					executionType: ExecutionType.HumanToAgent,
+					sessionId: conversationId,
+				},
+			};
+
+			const invokeAgentScope = InvokeAgentScope.start(invokeAgentDetails, tenantDetails);
+
+			await invokeAgentScope.withActiveSpanAsync(async () => {
+				invokeAgentScope.recordInputMessages([inputText || 'Unknown text']);
+
+				let microsoftMcpTools = undefined;
+				let mcpClient = undefined;
+				if (mcpTokenRef.token) {
+					try {
+						const result = await getMicrosoftMcpTools(turnContext, mcpTokenRef.token);
+						mcpClient = result?.client;
+						microsoftMcpTools = result?.tools;
+					} catch (error) {
+						console.log('Error retrieving MCP tools');
+					}
+				}
+
+				try {
+					const response = await invokeAgent(
+						nodeContext,
+						inputText,
+						agentDescription,
+						{
+							configurable: { thread_id: turnContext.activity.conversation!.id },
+						},
+						microsoftMcpTools,
+					);
+
+					invokeAgentScope.recordOutputMessages([`n8n Agent Response: ${response}`]);
+
+					await turnContext.sendActivity(response);
+				} finally {
+					if (mcpClient) await mcpClient.close();
+					invokeAgentScope.dispose();
+				}
+			});
+		});
+	};
+};
+
+export function configureAdapterProcessCallback(
+	nodeContext: IWebhookFunctions,
+	agent: AgentApplication<TurnState<DefaultConversationState, DefaultUserState>>,
+	credentials: MicrosoftAgent365Credentials,
+	activityCapture: ActivityCapture,
+) {
+	return async (turnContext: TurnContext) => {
+		const { token: aauToken } = await agent.authorization.exchangeToken(
+			turnContext,
+			getObservabilityAuthenticationScope(),
+			'agentic',
+		);
+
+		const observability = ObservabilityManager.configure((builder: Builder) =>
+			builder
+				.withService('TypeScript Sample Agent', '1.0.0')
+				.withTokenResolver((_agentId: string, _tenantId: string) => aauToken || ''),
+		);
+
+		observability.start();
+
+		const mcpTokenRef = { token: undefined as string | undefined };
+
+		try {
+			const tokenResult = await agent.authorization.getToken(turnContext, 'agentic');
+			mcpTokenRef.token = tokenResult.token;
```

Comment:
Doing some digging I've figured out that the second call to getToken here will just return a cached token that we fetch on line 189, and that token only contains observability scope.

What we should be doing is fetching the token with both scopes at once. Sadly, this has its own issue in agents-hosting package itself. I've logged an issue with them here: https://github.com/microsoft/Agents-for-js/issues/817


```suggestion
		const tokenResult = await agent.authorization.exchangeToken(turnContext, 'agentic', { scopes: [getMcpPlatformAuthenticationScope(), ...getObservabilityAuthenticationScope()] });

		const observability = ObservabilityManager.configure((builder: Builder) =>
			builder
				.withService('TypeScript Sample Agent', '1.0.0')
				.withTokenResolver((_agentId: string, _tenantId: string) => tokenResult.token || ''),
		);

		observability.start();

		const mcpTokenRef = { token: undefined as string | undefined };

		try {
			mcpTokenRef.token = tokenResult.token;
```

Reply:
Talking to some folks, this won't be supported because these are unrelated scopes. The fix then on the agents-hosting sdk side might be perhaps to improve the caching there to not return the previous token when scope changes. 

In the meanwhile I was testing how we can manipulate that cache to force a retrieval of the token a second time and this approach worked (both observability and mcp working):


```suggestion
		const { token: aauToken } = await agent.authorization.exchangeToken(
			turnContext,
			getObservabilityAuthenticationScope(),
			'agentic',
		);

		const observability = ObservabilityManager.configure((builder: Builder) =>
			builder
				.withService('TypeScript Sample Agent', '1.0.0')
				.withTokenResolver((_agentId: string, _tenantId: string) => aauToken || ''),
		);

		observability.start();

		const mcpTokenRef = { token: undefined as string | undefined };

		try {
			turnContext.turnState.set('AgenticAuthorization/agentic', undefined);
			const tokenResult = await agent.authorization.exchangeToken(turnContext, 'agentic', { scopes: [getMcpPlatformAuthenticationScope()] });
			mcpTokenRef.token = tokenResult.token;
```

Reply:
nice, thanks @rbrighenti :pray: